pytorch_lightning/utilities/auto_restart.py

```diff
@@ -476,6 +430,48 @@ def _capture_metadata_collate(
     return {"data": data, AutoRestartBatchKeys.PL_RESTART_META: metadata}
 
 
+# TODO: Merge this code within stateful DataLoaderIter.
+def _next_data_wrapper(fn, it, dl, num_batches_fetched, data_fetcher) -> Callable:
+    @wraps(fn)
+    def wrapper():
+        nonlocal num_batches_fetched
+        nonlocal it
+        nonlocal dl
+
+        dataset = dl.dataset
+        combined_batch = fn()
```

Comment:
I don't get why this is a wrapper if calling the passed fn is the first thing done inside.

Reply:
We are patching the iterator `_next_data`.

```
    iterator._next_data = _next_data_wrapper(
        iterator._next_data, iterator, dataloader, num_batches_fetched, data_fetcher
    )
```

With the refactor, this won't be necessary anymore as we will be subclassing the iterator directly and overriding _next_data.